pkg/reconciler/certificate/certificate.go

```diff
@@ -189,3 +201,37 @@ func (c *Reconciler) updateStatus(desired *v1alpha1.Certificate) (*v1alpha1.Cert
 
 	return c.ServingClientSet.NetworkingV1alpha1().Certificates(existing.Namespace).UpdateStatus(existing)
 }
+
+func (c *Reconciler) setHTTP01Challenges(knCert *v1alpha1.Certificate) error {
+	challenges := make([]v1alpha1.HTTP01Challenge, 0, len(knCert.Spec.DNSNames))
+	for _, dnsName := range knCert.Spec.DNSNames {
+		// This selector come from https://github.com/jetstack/cert-manager/blob/1b9b83a4b80068207b0a8070dadb0e760f5095f6/pkg/issuer/acme/http/pod.go#L34
+		selector := fmt.Sprintf("%s=%d", httpDomainLabel, adler32.Checksum([]byte(dnsName)))
+		svcs, err := c.KubeClientSet.CoreV1().Services(knCert.Namespace).List(metav1.ListOptions{LabelSelector: selector})
+		if err != nil {
+			return err
+		}
+		for _, svc := range svcs.Items {
+			owner := svc.GetOwnerReferences()[0]
+			cmChallenge, err := c.certManagerClient.CertmanagerV1alpha1().Challenges(knCert.Namespace).Get(owner.Name, metav1.GetOptions{})
```

Comment:
we can use certmanagerlisters.ChallengeLister to get challenges from cache, which will be more efficient.

Reply:
Will do.